pkg/controllers/dnspolicy/health.go

```diff
@@ -0,0 +1,217 @@
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthChecks(ctx context.Context, log logr.Logger, dnsRecord *v1alpha1.DNSRecord, policy *v1alpha1.DNSPolicy, config *healthChecksConfig) ([]dns.HealthCheckResult, error) {
+	healthCheckReconciler := r.DNSProvider.HealthCheckReconciler()
+
+	results := []dns.HealthCheckResult{}
+
+	for _, dnsEndpoint := range dnsRecord.Spec.Endpoints {
+		ok := false
+		if _, ok = dnsEndpoint.GetAddress(); !ok {
+			// TODO: Log
+			continue
+		}
+
+		endpointId, err := idForEndpoint(dnsRecord, dnsEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		spec := dns.HealthCheckSpec{
+			Id:               endpointId,
+			Name:             fmt.Sprintf("%s-%s", dnsEndpoint.DNSName, dnsEndpoint.SetIdentifier),
+			Path:             config.Endpoint,
+			Port:             config.Port,
+			Protocol:         config.Protocol,
+			FailureThreshold: config.FailureThreshold,
+		}
+
+		result, err := healthCheckReconciler.Reconcile(ctx, spec, dnsEndpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthCheckStatus(results []dns.HealthCheckResult, policy *v1alpha1.DNSPolicy) error {
+	for _, result := range results {
+		if result.Result == dns.HealthCheckNoop {
+			continue
+		}
+
+		if policy.Status.HealthCheck == nil {
+			policy.Status.HealthCheck = &v1alpha1.HealthCheckStatus{}
+		}
+
+		status := metav1.ConditionTrue
+		if result.Result == dns.HealthCheckFailed {
+			status = metav1.ConditionFalse
+		}
+
+		policy.Status.HealthCheck.Conditions = append(policy.Status.HealthCheck.Conditions, metav1.Condition{
+			ObservedGeneration: policy.Generation,
+			Status:             status,
+			Reason:             string(result.Result),
+			LastTransitionTime: metav1.Now(),
+			Message:            result.Message,
+			Type:               string(result.Result),
+		})
+	}
+
+	return nil
+}
+
+func (r *DNSPolicyReconciler) reconcileHealthCheckDeletion(ctx context.Context, dnsRecord *v1alpha1.DNSRecord, policy *v1alpha1.DNSPolicy) ([]dns.HealthCheckResult, error) {
+	reconciler := r.DNSProvider.HealthCheckReconciler()
+	results := []dns.HealthCheckResult{}
+
+	for _, endpoint := range dnsRecord.Status.Endpoints {
+		result, err := reconciler.Delete(ctx, endpoint)
+		if err != nil {
+			return nil, err
+		}
+
+		results = append(results, result)
+	}
+
+	return results, nil
+}
+
+func (r *DNSPolicyReconciler) getDNSRecord(ctx context.Context, policy *v1alpha1.DNSPolicy) (*v1alpha1.DNSRecord, bool, error) {
+	gateway, err := r.getGateway(ctx, policy)
+	if err != nil {
+		return nil, false, err
+	}
+
+	dnsRecordList := &v1alpha1.DNSRecordList{}
+	if err := r.Client.List(ctx, dnsRecordList); err != nil {
+		return nil, false, err
+	}
+
+	trafficAccessor := traffic.NewGateway(gateway)
+	allHosts := trafficAccessor.GetHosts()
+
+	if len(allHosts) == 0 {
+		return nil, false, nil
+	}
+
+	host := allHosts[0]
```

Comment:
why only the first one?

Reply:
My understanding is that it's a 1:1 relation between Gateways and DNSRecords, so there should be only one associated DNSRecord to the gateway. If I'm wrong then I'll update this to create a health check per DNSRecord

Reply:
no it is   a 1:1 releationship between a `hostname` and a dnsrecord and it is a 1:1 relationship between a specific subdomain and a gateway but a gateway can have many hostnames and so many DNSRecords

Reply:
I see. That makes more sense tbh. I'll make the changes